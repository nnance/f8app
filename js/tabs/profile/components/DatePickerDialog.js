import React from 'react';
import {
  DatePickerIOS,
  DatePickerAndroid,
  Modal,
  StyleSheet,
  View,
  Platform,
} from 'react-native';

import Button from 'Button';

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(0, 0, 0, 0.3)',
  },
  dialogContainer: {
    backgroundColor: 'rgba(255, 255, 255, 0.9)',
    borderRadius: 10,
    width: 300,
    borderWidth: 0.2,
    borderColor: 'rgba(0, 0, 0, 0.3)',
  },
  buttonContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
  },
  button: {
    flex: 1,
  },
});

class DatePickerDialog extends React.Component {

  constructor(...args) {
    super(...args);
    this.state = {
      currentDate: this.props.currentDate || new Date(),
      visible: false,
    };
    this.iResolve = null;
    this.iReject = null;
  }

  onOK() {
    if (this.iResolve) {
      this.iResolve(this.state.currentDate);
    }
    this.hide();
  }

  onCancel() {
    if (this.iReject) {
      this.iReject();
    }
    this.hide();
  }

  open(setDate) {
    const startDate = setDate || this.state.currentDate;
    if (Platform.OS === 'android') {
      return DatePickerAndroid.open({ date: startDate }).then((response) => {
        const { year, month, day } = response;
        return new Date(year, month, day);
      });
    }
    if (this.state.visible) {
      return this.iResolve;
    }
    this.setState({
      currentDate: startDate,
      visible: true,
    });
    return new Promise((resolve, reject) => {
      this.iResolve = resolve;
      this.iReject = reject;
    });
  }

  hide() {
    this.iResolve = null;
    this.iReject = null;
    this.setState({
      visible: false,
    });
  }

  render() {
    if (Platform.OS === 'android') {
      return null;
    }
    return (<Modal
      transparent
      visible={this.state.visible}
    >
      <View style={styles.container}>
        <View style={styles.dialogContainer}>
          <DatePickerIOS mode="date" date={this.state.currentDate} onDateChange={date => this.setState({ currentDate: date })} />
          <View style={styles.buttonContainer}>
            <View style={styles.button}>
              <Button title="ok" onPress={() => this.onOK()} />
            </View>
            <View style={styles.button}>
              <Button title="cancel" onPress={() => this.onCancel()} />
            </View>
          </View>
        </View>
      </View>
    </Modal>);
  }
}

export default DatePickerDialog;
