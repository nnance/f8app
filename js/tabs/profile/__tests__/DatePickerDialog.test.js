import React from 'react';
import renderer from 'react-test-renderer';
import Platform from 'Platform';
import { shallow } from 'enzyme';

import {
  DatePickerIOS,
  Button,
} from 'react-native';

/* eslint import/first: off */
jest.mock('DatePickerAndroid');
import DatePickerAndroid from 'DatePickerAndroid';
import DatePickerDialog from '../components/DatePickerDialog';

describe('DatePickerDialog', () => {
  let oldPlatform;

  beforeAll(() => {
    oldPlatform = Platform.OS;
  });
  afterAll(() => {
    Platform.OS = oldPlatform;
  });

  describe('android', () => {
    beforeAll(() => {
      Platform.OS = 'android';
    });

    it('render', () => {
      const tree = renderer.create(<DatePickerDialog currentDate={new Date(2016, 5, 12)} />);
      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('return picked date', async () => {
      let gResolve;
      const expectDate = new Date(2016, 5, 12);
      const openDate = new Promise((resolve) => {
        gResolve = resolve;
      });
      DatePickerAndroid.open.mockImplementation(() => openDate);
      const datePickerDialog = shallow(<DatePickerDialog />);
      gResolve({ action: undefined, year: 2016, month: 5, day: 12 });
      const date = await datePickerDialog.instance().open();
      expect(date.toString()).toBe(expectDate.toString());
    });
  });

  describe('ios', () => {
    beforeAll(() => {
      Platform.OS = 'ios';
    });

    it('render', () => {
      const tree = renderer.create(<DatePickerDialog currentDate={new Date(2016, 5, 12)} />);
      expect(tree.toJSON()).toMatchSnapshot();
    });

    it('return picked date', (done) => {
      const expectDate = new Date(2017, 5, 12);
      const wrapper = shallow(<DatePickerDialog />);
      const result = wrapper.instance().open(new Date(2016, 5, 12));
      result.then((date) => {
        expect(date).toBe(expectDate);
        done();
      });
      wrapper.find(DatePickerIOS).props().onDateChange(expectDate);
      wrapper.find(Button).forEach((button) => {
        if (button.props().title === 'ok') {
          button.simulate('press');
        }
      });
    });

    it('cancel pick', (done) => {
      const successFn = jest.fn();
      const wrapper = shallow(<DatePickerDialog />);
      const result = wrapper.instance().open();
      result.then(successFn).catch(() => {
        expect(successFn).not.toBeCalled();
        done();
      });
      wrapper.find(Button).forEach((button) => {
        if (button.props().title === 'cancel') {
          button.simulate('press');
        }
      });
    });
  });
});
