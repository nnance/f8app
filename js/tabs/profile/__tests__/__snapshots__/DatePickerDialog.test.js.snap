exports[`DatePickerDialog android render 1`] = `null`;

exports[`DatePickerDialog ios render 1`] = `
<Modal
  transparent={true}
  visible={false}>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "rgba(0, 0, 0, 0.3)",
        "flex": 1,
        "justifyContent": "center",
      }
    }>
    <View
      style={
        Object {
          "backgroundColor": "rgba(255, 255, 255, 0.9)",
          "borderColor": "rgba(0, 0, 0, 0.3)",
          "borderRadius": 10,
          "borderWidth": 0.2,
          "width": 300,
        }
      }>
      <View
        style={undefined}>
        <RCTDatePicker
          date={1465664400000}
          maximumDate={undefined}
          minimumDate={undefined}
          minuteInterval={undefined}
          mode="date"
          onChange={[Function]}
          style={
            Object {
              "height": 216,
            }
          }
          timeZoneOffsetInMinutes={undefined} />
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }>
        <View
          style={
            Object {
              "flex": 1,
            }
          }>
          <View
            accessibilityComponentType="button"
            accessibilityLabel={undefined}
            accessibilityTraits={
              Array [
                "button",
              ]
            }
            accessible={true}
            hitSlop={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID={undefined}>
            <View
              style={
                Array [
                  Object {},
                ]
              }>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "#0C42FD",
                      "fontSize": 18,
                      "padding": 8,
                      "textAlign": "center",
                    },
                  ]
                }>
                ok
              </Text>
            </View>
          </View>
        </View>
        <View
          style={
            Object {
              "flex": 1,
            }
          }>
          <View
            accessibilityComponentType="button"
            accessibilityLabel={undefined}
            accessibilityTraits={
              Array [
                "button",
              ]
            }
            accessible={true}
            hitSlop={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID={undefined}>
            <View
              style={
                Array [
                  Object {},
                ]
              }>
              <Text
                accessible={true}
                allowFontScaling={true}
                ellipsizeMode="tail"
                style={
                  Array [
                    Object {
                      "color": "#0C42FD",
                      "fontSize": 18,
                      "padding": 8,
                      "textAlign": "center",
                    },
                  ]
                }>
                cancel
              </Text>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</Modal>
`;
